xdsl/transforms/experimental/stencil_to_csl_stencil.py

```diff
@@ -0,0 +1,163 @@
+from xdsl.dialects.builtin import IntegerAttr, IntegerType, ModuleOp, TensorType
+from xdsl.dialects.csl_ import csl_stencil
+from xdsl.dialects.experimental import dmp
+from xdsl.ir import Attribute, OpResult
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.hints import isa
+
+
+@dataclass(frozen=True)
+class ApplyOpAddPrefetch(RewritePattern):
+
+    prefetch_op: csl_stencil.PrefetchOp
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.ApplyOp, rewriter: PatternRewriter, /):
+
+        pass
+
+
+@dataclass(frozen=True)
+class AccessOpFromPrefetch(RewritePattern):
+
+    # index of the stencil access to be replaced by the prefetched buffer in the last arg
+    arg_index: int
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.AccessOp, rewriter: PatternRewriter, /):
+        assert len(op.offset) == 2
+        if tuple(op.offset) == (0, 0):
+            return
+        prefetched_arg = op.get_apply().region.block.args[-1]
+        assert isa(m_type := prefetched_arg.type, memref.MemRefType[Attribute])
+        assert isa(t_type := m_type.get_element_type(), TensorType[Attribute])
+
+        csl_access_op = csl_stencil.AccessOp(
+            op=prefetched_arg,
+            offset=op.offset,
+            offset_mapping=op.offset_mapping,
+            result_type=t_type,
+        )
+
+        # if the matched op is immediately followed by tensor.ExtractSliceOps to remove ghost cells, replace both
+        if len(op.res.uses) == 1 and isinstance(
+            use := list(op.res.uses)[0].operation, tensor.ExtractSliceOp
+        ):
+            rewriter.replace_op(use, csl_access_op)
+            rewriter.erase_op(op)
+        else:
+            rewriter.replace_matched_op(csl_access_op)
+
+
+@dataclass(frozen=True)
+class SwapToPrefetch(RewritePattern):
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: dmp.SwapOp, rewriter: PatternRewriter, /):
+        if op.swaps is None or len(op.swaps) == 0:
+            rewriter.erase_matched_op(False)
+            return
+        assert isinstance(op.input_stencil, OpResult)
+        # currently only works for 3-dimensional stencils
+        assert all(len(swap.size) == 3 for swap in op.swaps)
+        # dmp should decompose from (x,y,z) to (1,1,z)
+        assert all(swap.size[:2] == (1, 1) for swap in op.swaps)
+        # check that size is uniform
+        uniform_size = op.swaps.data[0].size[2]
+        assert all(swap.size[2] == uniform_size for swap in op.swaps)
+
+        assert isa(
+            op.input_stencil.type,
+            memref.MemRefType[Attribute] | stencil.TempType[Attribute],
+        )
+        assert isa(
+            t_type := op.input_stencil.type.get_element_type(), TensorType[Attribute]
+        )
```

Comment:
I really don't know how I feel about all these asserts in the rewrites... @PapyChacal @superlopuh do you know of a better way to encode specific requirements of the input IR on the pass?

I would at least like to see these asserts produce a meaningful warning about what is allowed and what isn't...